### PR TITLE
FileSink: close an ended sink that still drains when its VM is torn down

### PR DESCRIPTION
**Pull request by robobun**
### Problem
- A Worker ends after a `type: "direct"` stream into a `Bun.spawn` stdin ended, before the child drained the pipe. One FileSink leaks with the pipe open, and the child, which outlives the worker, never sees EOF. Linux, main and canary.
- The controller's `end()`/`close()` (`__endImpl`, `__closeWithReason` in `src/codegen/generate-jssink.ts`) detaches the controller before it ends the sink. If the flush leaves bytes (`Pending` arm of `FileSink::end`/`end_from_js`), only the keep-alive ref holds it. POSIX teardown never delivers the `on_close` that releases it.

### Fix
- An ended sink that still drains joins its VM's root context through the `AbortHandle` it embeds (`close_with_vm`). The teardown stop phase closes the writer, and `on_close` releases the ref.
- It joins only while `VirtualMachine::stop_phase_has_begun()` is false. `teardown` sets that flag before its first sweep, so each joined sink is stopped.
- The handler ignores a `bun test --isolate` swap, whose loop still drains the write. Not on Windows: its stop phase already closes the worker's pipes.
- Verified: new test in `test/js/web/workers/worker-terminate-lifetime.test.ts`, which fails with main's `src/`. Self-reviewed: 6 concerns, 4 addressed, 2 rejected (Notes).

### Background
- A `FileSink` is refcounted. A write the fd cannot take at once takes a keep-alive ref until the writer closes.
- A `ScriptExecutionContext` lists the `AbortHandle`s of native resources. VM teardown stops each.

### Downsides
- A POSIX sink that ended with buffered bytes is now closed at VM teardown, and the bytes are dropped. They were never written before either.
- Cost: a `Cell<bool>` per FileSink and per VM, one list insert per `end()` that cannot flush.

<details><summary>Notes</summary>

**Measured.** With main's `src/` the new test prints `{"leakedFileSinks":2,"childrenThatSawEOF":0}`, on a debug ASAN build and on the canary. `fileSinkInternals.liveCount()` stays +1 per sink after the `exit` event of the worker, on main and on canary `1.4.3-canary.1+367d939d9`.

**Related.** Related to #43729 (open): that PR handles a controller that dies still attached to its sink. Here the controller is already detached, so this leak survives it, and its Notes hand this case off. #43761 (open) adds a guard at the top of `FileSink::on_close`, one line above the line this PR adds there. Whichever lands second has a mechanical conflict: keep both.

**Does the use-after-free of #43761 reach the new POSIX path?** No. The abort handler takes its own ref before `writer.close()`, so `Writable::on_close` dropping the ref of the Subprocess inside `on_close` cannot free the sink.

**Repro.** A worker runs `Bun.spawn({ stdin: new ReadableStream({ type: "direct", async pull(c) { c.write(new Uint8Array(4 << 20)); await 1; c.end(); } }) })` with a child that does not read yet. It reports, and the parent calls `terminate()`.

**Same leak, other doors (canary leaks 1, fixed build 0).** `controller.close()` in place of `end()`. A synchronous `pull`. A worker that ends with `process.exit()`. `Bun.write(fifo, new Response(directStream))` in a worker, where the reader of the FIFO then sees EOF.

**Why Windows is gated off.** An ungated build regressed it. A worker that does `Bun.file(p).writer()`, writes 64 MB, calls `end()` and exits wrote 1 MB of 64 MB, where main writes all 64 MB. On Windows the teardown drains the requests of the loop, so an ended file sink still writes out its buffer. With the gate it writes 64 MB again. POSIX has no such case: a regular file never returns `Pending` (only `EAGAIN` does), and nothing ticks the loop once teardown starts.

**The arming guard.** A sink joins the root context only while a stop sweep is still to come, or nothing would unlink it. `VirtualMachine::stop_phase_has_begun()` answers that exactly: `teardown` sets the flag right before its first sweep. Arming during a sweep is refused, so the `second == SweepResult::Idle` debug assertion in `teardown` cannot trip. The first push of this PR guessed the answer from `is_shutting_down` and `script_allowed()`. `script_allowed()` is cleared by a parent's `terminate()` from the parent's thread at request time, long before the worker's sweep. That guess refused a sink that ended in the exit-time flush of a worker its parent terminated, although the sweep follows. A review comment on this PR pointed that out, and the flag replaces the guess.

**What no test reaches.** The new test terminates the worker only after the stream ended and its reactions ran. No deterministic test covers an `end()` that lands between a terminate request and the sweep: after the request, JS throws at its next trap check, so script cannot call `end()` on demand inside that window. The flag covers that window by construction. The existing test and the probes below exercise the same arming path.

**My own review before the first push: addressed.**
1. The ASAN CI lanes set `BUN_FEATURE_FLAG_NO_ORPHANS`. A child that inherits it is SIGKILLed when the thread that spawned it exits, which is the worker this test terminates. Reproduced locally as a 60 s timeout with the fix in place. The test now unsets it for its process tree.
2. A guard on `script_allowed()` alone skipped an `end()` that raced a terminate request. See the arming guard above.
3. Whether the leak detection and exception validation of that lane stay quiet. The children inherit them, and the test asserts that their stderr is empty. Ran the whole file under that environment: 26 of 26 pass.
4. The pointer the handle is armed with. See the next section. I rejected this at first, and that was wrong.

**My own review: rejected.**
1. A `Bun.ModuleGraph` script that calls `start({ path })` again on a sink still draining in the root context would not join the graph. The non-pollable form of that restart already left its graph before this change.
2. Reporting whether `write()` hit backpressure, to catch a silent pass on a host whose `net.core.wmem_default` exceeds 4 MiB. The CI images do not raise it, and the check would need a POSIX gate.

**Changed after review on this PR.**
1. The abort handler can free the sink through the pointer it was armed with. Both arm sites (`close_with_graph`, which predates this PR, and the new `close_with_vm`) passed `from_ref(self).cast_mut()`. The first version of these Notes said this PR added "the same caveat and no new one". That was not accurate: it added a second instance of a pattern `src/CLAUDE.md` forbids, on a path where the handler is usually the one that frees. Both sites now pass the allocation's own pointer, which the writer holds for every sink that `init` or `create*` made (`abort_owner`). A sink built by the JS constructor never gets that pointer (nothing calls `set_parent` for it), so it alone keeps the old pointer. Giving it one means a hook in the generic sink constructor, which is a separate change.
2. The arming guard, as described above.
3. The multi-line comments this PR added under `src/` are now one line each, as the repository's comment check asks. The reasons moved here and into the commit messages.
4. The test fixture awaited the worker's message and the EOF reports of the children with no failure path, so a worker that died early or a child that went away made the test wait for its timeout. `REVIEW.md` asks that every failure event rejects the awaited promise. The worker's `error` and early `exit` now reject the wait, as the other tests in this file do. A child reports a read that threw. For a child that dies without running any JS, the parent has no process handle (the worker spawned it, and the worker is gone), so the gate response stays open after its answer and the abort of that request is the exit of the child. There is no deadline and no timer. I ran each of the four paths once with a sabotaged fixture that is not committed: each ends in 2 to 3 s on the debug build with its reason on stderr, where it waited 60 s before.

**`bun test --isolate`.** A probe with an ended, draining stdin sink left across a file swap shows the same FileSink counts before and after this change.

**Suites run on the Linux debug ASAN build with the fix.** `worker-terminate-lifetime.test.ts` (26 of 26 under the environment of the ASAN lane), `filesink.test.ts`, `spawn-stdin-pipe-fd-leak`, `spawn-streaming-stdin`, `spawn-stdin-destroy`, `spawn-pipe-start-error`, `blob-write`, `console-write`, `serve-direct-readable-stream`, and the three FileSink tests of `module-graph-isolation`. The new test also passed 8 of 8 under the lane's environment after its last change. On windows-x64 the new test passes with and without the fix. The pointer and guard change was built and tested there too, and a worker's file sink still writes all 64 MB. The final state of this branch was built there from the same diff, and the new test passes 4 of 4.

**Local failures that do not come from this change.** Two 5 s timeouts in `spawn-stdin-readable-stream.test.ts` and one in its edge-cases file: a debug build of main's `src/` shows the same three. `bun-write.test.js` "Bun.file(0) survives GC" times out at 5 s, and it does not touch this path. `module-graph-isolation` shows 5 s timeouts under load in tests unrelated to FileSink, and the ones rerun alone pass.

**Not changed.** A default (non-direct) stream still parked on backpressure at `terminate()` hits `ASSERTION FAILED: decontaminate()` on a debug build. That is the crash #43729 reports on main. This change does not arm that sink, because it has not ended.

</details>

---

**no test proof** · iteration 0 · platform-specific test(s) that do not run on this machine, deferring to CI, which covers all platforms: test/js/web/workers/worker-terminate-lifetime.test.ts

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 10:51 PM PT - Sep 21st, 2026</sup></div>

@robobun, your commit ff984b43ccde1f32db7373faf67eb53fa4a1d238 is building: [`#119601`](https://buildkite.com/bun/bun/builds/119601)

**Comment by robobun** [comment]
**Status**

How I reproduced it. A worker calls `Bun.spawn` with `stdin: new ReadableStream({ type: "direct", async pull(c) { c.write(new Uint8Array(4 << 20)); await 1; c.end(); } })`. The child does not read yet. The worker reports, and the parent calls `terminate()`.

- With main's `src/`: `fileSinkInternals.liveCount()` stays +1 after the `exit` event of the worker, and the child never sees EOF on stdin. The new test prints `{"leakedFileSinks":2,"childrenThatSawEOF":0}` (one cell for `controller.end()`, one for `controller.close()`). This is the same on a debug ASAN build and on canary `1.4.3-canary.1+367d939d9`.
- With this change: `{"leakedFileSinks":0,"childrenThatSawEOF":2}`.
- Windows does not have the bug. The test passes there with and without the change.

Run it with `bun bd test test/js/web/workers/worker-terminate-lifetime.test.ts -t "direct stream ended"`.

This PR: #43769. Related: #43729, #43761.

File: src/runtime/webcore/FileSink.rs [hunk 1]
```diff
@@ -504,6 +507,34 @@ impl FileSink {
...
+            if !stop_phase_is_to_come || self.abort_handle.context_id().is_some() {
+                return;
+            }
+            // SAFETY: as in `close_with_graph`.
+            unsafe {
+                bun_jsc::AbortHandle::arm_owner(
+                    core::ptr::from_ref(self).cast_mut(),
+                    vm.root_context(),
+                );
+            }
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/FileSink.rs]: You are right, and the bound in my Notes ("the same caveat and no new one") was not accurate. This PR added a second instance of the pattern, on a path where the handler is usually the one that frees the sink.

Changed in 365d202e97. Both arm sites (`close_with_graph` and `close_with_vm`) now pass the allocation's own pointer through a new `abort_owner()`. The SAFETY comment of the handler no longer makes the blanket provenance claim. It points at `abort_owner`.

One correction: that pointer is not available for every sink. A sink built by the JS constructor (`js_construct` in `Sink.rs`) never has `set_parent` called for its writer, so its `parent` is null. Arming with it would push a bad pointer into the context list. `abort_owner()` checks for null, and only that sink keeps the old `&self` pointer. To give it a real one needs a hook in the generic sink constructor, which is a separate change. The Notes now say this.

File: src/runtime/webcore/FileSink.rs [hunk 2]
```diff
@@ -504,6 +507,34 @@ impl FileSink {
...
+        #[cfg(not(windows))]
+        {
+            let Some(vm) = self.js_vm() else { return };
+            // The stop phase is still to come while this thread has not begun its shutdown, or
+            // while script may run (teardown forbids it before its first sweep). A parent's
+            // `terminate()` forbids script from its own thread at once, so that alone proves
+            // nothing. Past both, the sweeps can be over and nothing would unlink this sink.
+            let stop_phase_is_to_come = !vm.is_shutting_down() || vm.script_allowed();
+            if !stop_phase_is_to_come || self.abort_handle.context_id().is_some() {
+                return;
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/FileSink.rs]: Agreed, the guard was stricter than it had to be. Changed in 365d202e97 as you suggest.

`VirtualMachine` has a new `stop_phase_begun: Cell<bool>`. `teardown` sets it right before its first sweep, and `close_with_vm` asks `vm.stop_phase_has_begun()`. The guess from `is_shutting_down` and `script_allowed()` is gone. A sink that ends in the exit-time flush of a parent-terminated worker now joins the root context, and the sweep that follows closes it. A sink that would join during or after a sweep is still refused, so the `second == SweepResult::Idle` assertion cannot trip.

It is a `Cell` so that `teardown` can set it through the shared `&VirtualMachine` it already holds, like the two flags above it.

No deterministic test reaches that window. I removed the matching item from Downsides, and the Notes say what the tests do and do not cover. The whole test file passes 26 of 26 under the environment of the ASAN lane.

File: src/jsc/VirtualMachine.rs [hunk 3]
```diff
@@ -345,6 +345,9 @@ pub struct VirtualMachine {
     /// When true, drainMicrotasksWithGlobal is suppressed. `Cell` for the same
     /// reason as [`Self::is_inside_deferred_task_queue`].
     pub(crate) suppress_microtask_drain: core::cell::Cell<bool>,
+    /// [`teardown`](Self::teardown) is about to run its first stop-phase sweep, or has: what joins
+    /// a context from here is not stopped. `Cell` and zero-valid as the two above.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/VirtualMachine.rs]: Reduced to one line in 9496c071bf.

File: src/jsc/VirtualMachine.rs [hunk 4]
```diff
@@ -1709,6 +1712,13 @@ impl VirtualMachine {
         self.is_shutting_down
     }
 
+    /// An owner that joins a context only so that teardown stops it asks this first: once the
+    /// sweeps have started nothing would unlink it. `script_allowed()` is no substitute, because a
+    /// parent's `terminate()` forbids script from its own thread long before this thread's sweep.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/VirtualMachine.rs]: Reduced to one line in 9496c071bf. It keeps the one fact that is easy to get wrong: `script_allowed()` is not a substitute.

File: src/runtime/webcore/FileSink.rs [hunk 5]
```diff
@@ -66,8 +66,11 @@ pub struct FileSink {
     /// while an async operation is pending. This is set when endFromJS returns a
     /// pending Promise and cleared when the operation completes.
     pub(crate) js_sink_ref: JsCell<bun_jsc::strong::Optional>,
-    /// Armed while a file this sink opened for a `Bun.ModuleGraph`'s script is open.
+    /// Armed while a file this sink opened for a `Bun.ModuleGraph`'s script is open, and while
+    /// an ended sink drains (`close_with_vm`).
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/FileSink.rs]: Reduced to one line in 9496c071bf.

File: src/runtime/webcore/FileSink.rs [hunk 6]
```diff
@@ -498,9 +501,38 @@ impl FileSink {
...
             // SAFETY: a started sink is heap-allocated; it leaves its context in `on_close`.
-            unsafe {
-                bun_jsc::AbortHandle::arm_owner(core::ptr::from_ref(self).cast_mut(), context);
+            unsafe { bun_jsc::AbortHandle::arm_owner(self.abort_owner(), context) };
+        }
+    }
+
+    /// What `abort_handle` is armed with. Its handler can free this sink through the pointer, so
+    /// it is the allocation's own (`init`/`create*` gave it to the writer) and not one derived
+    /// from `&self`. Only a sink the JS constructor built has no such pointer.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/FileSink.rs]: Reduced to one line in 9496c071bf. The null fallback stays in the code: a sink built by the JS constructor never gets the pointer, and to give it one is a separate change in the generic sink constructor.

File: src/runtime/webcore/FileSink.rs [hunk 7]
```diff
@@ -498,9 +501,38 @@ impl FileSink {
...
+        } else {
+            allocation
+        }
+    }
+
+    /// This sink ended with bytes still buffered. From here only the writer's close releases the
+    /// keep-alive ref, and no JS cell may be left that reaches the sink (a controller detaches
+    /// before it ends it). A VM torn down before the drain never delivers that close, so its
+    /// stop phase closes the writer instead. Not on Windows: that phase closes every pipe on the
+    /// worker's loop there, and a file's buffered writes finish while the loop's requests drain.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/FileSink.rs]: Reduced to one line in 9496c071bf. The reason for the function moved to the commit message and the PR description. The line that stays says why Windows is gated off.

**Comment by robobun** [comment]
Both review findings were right, and both are addressed in 365d202e97. The replies in the two threads have the details.

- **Pointer the abort handle is armed with.** Both arm sites now pass the allocation's own pointer (`abort_owner`). My Notes had called this "the same caveat and no new one". That was not accurate, and the Notes now say so. One limit remains: a sink built by the JS constructor never gets that pointer, so it alone keeps the old one.
- **Arming guard.** `VirtualMachine::stop_phase_has_begun()` replaces the guess from `is_shutting_down` and `script_allowed()`. `teardown` sets the flag right before its first sweep. The case that the Downsides listed as still leaking is covered now, so I removed it from there. No deterministic test reaches that window, and the Notes say that.

9496c071bf brings the comments this PR adds under `src/` down to one line each, as the comment check asks. The check found nothing new on that commit.

Verified on the Linux debug ASAN build: the new test, the whole test file under the environment of the ASAN lane (26 of 26), and the FileSink suites. 365d202e97 was also built and tested on windows-x64, where a worker's file sink still writes all 64 MB. 9496c071bf changes comments only.

File: test/js/web/workers/worker-terminate-lifetime.test.ts [hunk 8]
```diff
@@ -960,6 +960,109 @@ test(
...
+            return new Response(await workerGone[cell].promise);
+          },
+        });
+        (async () => {
+          const baseline = fileSinkInternals.liveCount();
+          await Promise.all(cells.map(async (how, cell) => {
+            const w = new Worker(join(__dirname, "worker.js"), { workerData: { url: server.url.href, cell, how } });
+            // The exit event follows the worker's VM teardown.
+            const exited = new Promise(resolve => w.once("exit", resolve));
+            await new Promise(resolve => w.once("message", resolve));
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/web/workers/worker-terminate-lifetime.test.ts]: Agreed. `REVIEW.md` asks for this ("wire EVERY failure event to reject the awaited promise"), and I had accepted a wait for the timeout as the failure mode. Changed in 4f3fff46e9.

- The worker wait uses the same helper as lines 304-306: `error` and an early `exit` reject it.
- A child reports a read that threw (`failed/<cell>`), and the EOF wait rejects with that text.
- For a child that dies without running any JS, the parent has no process handle, because the worker spawned it and the worker is gone. So the gate response now stays open after its answer, and the abort of that request is the exit of the child. No deadline and no timer.
- `main.js` prints the reason and exits 1.

I ran each path once with a sabotaged fixture (not committed). All four end in 2 to 3 s on the debug build, where they waited 60 s before:
- a throw in the worker: `error: sabotaged worker`
- an early worker exit: `error: worker exited 7 before its stream ended`
- a throw in the read of the child: `error: child 0 failed: Error: sabotaged read`
- SIGKILL of the child: `error: child 0 exited before it reported EOF`

The two main paths did not change: main's `src/` still gives `{"leakedFileSinks":2,"childrenThatSawEOF":0}` with an empty stderr, and the fix passes. It passes 8 of 8 under the environment of the ASAN lane and 4 of 4 on windows-x64.

**Comment by robobun** [comment]
The test nit is addressed in 4f3fff46e9, and the reply in its thread has the details. A worker or a child that fails now ends the test in 2 to 3 s with its reason on stderr, where it waited for the timeout before. There is no deadline and no timer: the worker's `error` and early `exit` reject the wait, and the exit of a child arrives as the abort of its gate request, which stays open.

The two main paths did not change. It passes 8 of 8 under the environment of the ASAN lane and 4 of 4 on windows-x64.
